docs: Immutable static assets (#95348)
vercel/next.js: docs/01-app/03-api-reference/05-config/01-next-config-js/supportsImmutableAssets.mdx

---
title: supportsImmutableAssets

> description: Enable Immutable Static Assets support

Reviewer: ```suggestion
description: Configure support for immutable static assets
```

---

> **Attention**: This flag is mostly relevant for adapter authors.
> Users of Next.js should only set this flag if they are encountering issues with their adapter.
>

> > **Enabling this feature even though your provider/adapter does not support it may result in broken deployments.**

Reviewer: ```suggestion
> **Attention**: This option is primarily intended for [adapter](/docs/app/api-reference/adapters) authors.
> App developers should only set it if instructed by their deployment provider or adapter maintainer, or when troubleshooting adapter-specific issues.
>
> **Enabling this feature when your provider or adapter does not support it can result in broken deployments.**
```

A bit unsure about the second paragraph. I am trying to make it clear that "on your own" you don't need to use this flag, it'll come to you, so to speak.

> Enabling skew protection means that all static assets are requested with a deployment-specific query parameter, e.g.

Reviewer: ```suggestion
When skew protection is enabled, static asset requests include a deployment-specific query parameter, for example:
```

```plain
GET https://foo.com/_next/static/chunks/0d_ks0ow7ur6m.js?dpl=<unique-deployment-id>
```

> One downside of this is that browsers have to download even unchanged static assets again after a new deployment.

Reviewer: ```suggestion
One downside is that browsers have to download static assets again after each new deployment, even when those assets have not changed.
```

> Immutable static assets is a feature that enables omitting the `?dpl` query parameter for static assets that are guaranteed to be immutable and content-addressed by their filename, allowing browsers to cache them indefinitely. It also enables not uploading unchanged static assets again after a new deployment:

Reviewer: ```suggestion
Immutable static assets allow the `?dpl` query parameter to be omitted for static assets that are guaranteed to be immutable and content-addressed by their filename. This lets browsers cache those assets indefinitely and allows unchanged static assets to be skipped when uploading files during subsequent deployments:
```

```plain
GET https://foo.com/_next/static/immutable/chunks/0d_ks0ow7ur6m.js
```

> You can use this config option to opt out of immutable static assets (when your adapter enabled it, otherwise this has no effect):

Reviewer: ```suggestion
You can use this config option to opt out of immutable static assets when your adapter has enabled support for them. If your adapter has not enabled this feature, this option has no effect:
```

```js filename="next.config.js"
/** @type {import('next').NextConfig} */
const nextConfig = {
  supportsImmutableAssets: false,
}

module.exports = nextConfig
```

See [Supporting immutable static assets](/docs/app/api-reference/adapters/immutable-static-assets) for more information on how to support this feature in an adapter.

## Version History

| Version   | Changes                                                |
| --------- | ------------------------------------------------------ |

> | `v16.3.0` | Support for immutable static assets feature was added. |

Reviewer: `Added support for immutable static assets.`

```suggestion
| `v16.3.0` | Added support for immutable static assets. |
```
